batch_requests.go

```diff
@@ -0,0 +1,210 @@
+				return res, err
+			}
+			err = json.Unmarshal(jsonStr, &res)
+			if err != nil {
+				return res, err
+			}
+
+			return res, nil
+		}
+	}
+
+	return res, errors.New("Response not found, check if id is valid")
+}
+
+// NewBatchRequest creates an instance of BatchRequest
+func NewBatchRequest() *batchRequest {
+	return &batchRequest{}
+}
+
+func toBatchItem(requestInfo abstractions.RequestInformation) (*batchItem, error) {
+	uri, err := requestInfo.GetUri()
+	if err != nil {
+		return nil, err
+	}
+
+	var body map[string]any
+	err = json.Unmarshal(requestInfo.Content, &body)
+	if err != nil {
+		return nil, err
+	}
+
+	return &batchItem{
+		Id:        uuid.NewString(),
+		Method:    requestInfo.Method.String(),
+		Body:      body,
+		Headers:   requestInfo.Headers,
+		Url:       uri.Path,
+		DependsOn: make([]string, 0),
+	}, nil
+}
+
+func (br *batchRequest) toJson() ([]byte, error) {
+	return json.Marshal(br)
+}
+
+func getBaseUrl(adapter abstractions.RequestAdapter) (*url.URL, error) {
+	return url.Parse(adapter.GetBaseUrl())
+}
+
+func buildRequestInfo(jsonBody []byte, baseUrl *url.URL) *abstractions.RequestInformation {
+	requestInfo := abstractions.NewRequestInformation()
+
+	requestInfo.SetStreamContent(jsonBody)
+	requestInfo.Method = abstractions.POST
+	requestInfo.SetUri(*baseUrl)
+	requestInfo.Headers = map[string]string{
+		"Content-Type": "application/json",
+	}
+
+	return requestInfo
+}
+
+func sendBatchRequest(requestInfo *abstractions.RequestInformation, adapter abstractions.RequestAdapter) (*BatchResponse, error) {
+	if requestInfo == nil {
+		return nil, errors.New("requestInfo cannot be nil")
+	}
+
+	// SendAsync type asserts HandlerFunc's return value into a parsable. We bypass SendAsync deserialization by returning a noop
+	// parsable struct and directly marshalling the response into a struct.
+	var res BatchResponse
+	_, err := adapter.SendAsync(requestInfo, nil, func(response any, errorMappings abstractions.ErrorMappings) (any, error) {
+		resp, ok := response.(*http.Response)
+		if !ok {
+			return nil, errors.New("Response type assertion failed")
+		}
+
+		if status := resp.StatusCode; status >= 400 && status < 600 {
+			return nil, fmt.Errorf("Request failed with status: %d", status)
+		}
+
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		err = json.Unmarshal(body, &res)
+		if err != nil {
+			return nil, err
+		}
+
+		return internal.NewNoOpParsable(), nil
+	}, nil)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &res, nil
+}
+
+type batchRequest struct {
+	Requests []*batchItem `json:"requests"`
+}
+
+type BatchResponse struct {
+	Responses []batchItem
+}
+
+type errorDetails struct {
+	Code    string
+	Message string
+}
+
+func (e errorDetails) Error() string {
+	return fmt.Sprintf("Code: %s \n Message: %s", e.Code, e.Message)
+}
+
+type errorResponse struct {
+	Error errorDetails
+}
+
+type batchItem struct {
+	Id        string            `json:"id"`
```

Comment:
I do think we should try to make it a model (implementing Parsable) so we can leverage the serialization infrastructure we already have in place. Same for the response.